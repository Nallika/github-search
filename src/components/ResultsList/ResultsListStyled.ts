import styled, {keyframes} from 'styled-components';

const spin = keyframes`
  to {
    transform: rotate(360deg);
  }
`;

const LoaderContainer = styled.div`
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`;

const BottomLoaderContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100px;
`;

const Loader = styled.div`
  display: inline-block;
  width: 40px;
  height: 40px;
  border: 4px solid gray;
  border-top: 4px solid green;
  border-radius: 50%;
  animation: ${spin} 1s linear infinite;
`;

const Wrapper = styled.section`
  margin: 5px;
  padding-top: 10px;
  height: 100%;
`;

const Text = styled.p`
  font-size: medium;
  font-weight: bold;
  padding-bottom: 10px;
  margin: 0;
`

export { Wrapper, Loader, LoaderContainer, BottomLoaderContainer, Text }
