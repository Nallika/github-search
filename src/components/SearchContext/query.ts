import { gql } from '@apollo/client';

export const GET_REPOS = gql`
  query($query: String!, $first: Int!, $after: String) {
    search(query: $query, type: REPOSITORY, first: $first, after: $after) {
      repositoryCount
      pageInfo {
        endCursor
      }
      edges {
        node {
          ... on Repository {
            id
            name
            stargazerCount
            forkCount
            url
          }
        }
      }
    }
  }
`;
