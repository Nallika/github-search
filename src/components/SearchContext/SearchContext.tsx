import React, {createContext, useCallback, useContext, useState} from 'react';
import { useLazyQuery } from '@apollo/client';

import { SearchContextValues, SearchProviderProps } from '../../types';
import { GET_REPOS } from './query';
import { DEFAULT_OFFSET } from '../../constants';

const SearchContext = createContext<SearchContextValues>({
  results: [],
  count: 0,
  loading: false,
  requestSearch: () => {}
});

export const SearchProvider: React.FC<SearchProviderProps> = ({ children }) => {
  const [getData, { loading, data }] = useLazyQuery(GET_REPOS);
  const [ searchQuery, setSearchQuery ] = useState('');
  const [ searchResults, setSearchResult ] = useState([]);
  const [ resultsCount, setResultsCount ] = useState(0);

  const requestSearch = useCallback(async (query: any) => {
    if (loading) {
      return;
    }

    const isNewSearch = query !== searchQuery;

    try {
      const offset = isNewSearch ? null : data.search.pageInfo.endCursor;

      const { data: { search: {edges: results, repositoryCount}} } = await getData({
        variables: {
          query,
          first: DEFAULT_OFFSET,
          after: offset
        }
      });

      setResultsCount(repositoryCount);
      setSearchQuery(query);
      setSearchResult(results);

    } catch (error) {
      console.error(error);
    }
  }, [data, getData, loading, searchQuery]);

  return (
    <SearchContext.Provider value={{
      results: searchResults,
      count: resultsCount,
      loading,
      requestSearch
    }}>
      {children}
    </SearchContext.Provider>
  );
};

export const useSearch = () => useContext(SearchContext);
