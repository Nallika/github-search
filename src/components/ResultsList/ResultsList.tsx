import React from 'react';

import { useSearch } from '../SearchContext';
import { SearchItem } from '../../types';
import ResultItem from '../ResultItem';
import { Wrapper, Loader, CountText } from './ResultsListStyled';

/**
 *
 * @constructor
 */
const ResultsList = () => {
  const { results, count, loading } = useSearch();
  const showList = !loading && results?.length;

  return (
    <Wrapper>
      {loading && <Loader />}
      {showList ?
        <>
          <CountText>{`Found ${count} repositories`}</CountText>
          {results.map((item: SearchItem) => (
            <ResultItem {...item.node} />
          ))}
        </> : null
      }
    </Wrapper>
  );
}

export default ResultsList;
