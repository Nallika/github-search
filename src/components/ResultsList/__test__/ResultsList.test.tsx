import React from 'react';
import { render, screen } from '@testing-library/react';
import ResultsList from '../ResultsList';
import * as SearchContext from '../../SearchContext';
import {useInView} from "react-intersection-observer";

jest.mock('../../SearchContext', () => ({
  ...jest.requireActual('../../SearchContext'),
  useSearch: jest.fn(),
}));

jest.mock('react-intersection-observer', () => ({
  useInView: jest.fn(),
}));

describe('ResultsList', () => {
  const searchMock = {
    results: [],
    count: 0,
    loading: true,
    isEmpty: false,
    requestMore: jest.fn(),
  }

  it('`<ResultsList />` Should render component', () => {
    (SearchContext.useSearch as jest.Mock).mockReturnValue(searchMock);

    (useInView as jest.Mock).mockReturnValue([jest.fn(), false]);

    const { container } = render(<ResultsList />);

    expect(container).toMatchSnapshot();
  });

  it('renders loading spinner when loading and no results', () => {
    (SearchContext.useSearch as jest.Mock).mockReturnValue(searchMock);

    (useInView as jest.Mock).mockReturnValue([jest.fn(), false]);

    render(<ResultsList />);

    expect(screen.getByTestId('loader')).toBeInTheDocument();
  });

  it('renders results and "Found X repositories" text when results exist', () => {
    const mockResults = [
      { node: { id: '1', name: 'Repo1', stargazerCount: 10 } },
      { node: { id: '2', name: 'Repo2', stargazerCount: 20 } },
    ];

    (SearchContext.useSearch as jest.Mock).mockReturnValue({
      ...searchMock,
      results: mockResults,
      count: mockResults.length,
    });

    (useInView as jest.Mock).mockReturnValue([jest.fn(), false]);

    render(<ResultsList />);

    expect(screen.getByText('Found 2 repositories')).toBeInTheDocument();

    mockResults.forEach((result) => {
      expect(screen.getByText(result.node.name)).toBeInTheDocument();
      expect(screen.getByText(`Stars: ${result.node.stargazerCount}`)).toBeInTheDocument();
    });

    expect(screen.queryByTestId('loader')).not.toBeInTheDocument();
  });

  it('calls requestMore when in view', async () => {
    const mockResults = [
      { node: { id: '1', name: 'Repo1', stargazerCount: 10 } },
      { node: { id: '2', name: 'Repo2', stargazerCount: 20 } },
    ];
    const mockRequestMore = jest.fn();

    (SearchContext.useSearch as jest.Mock).mockReturnValue({
      results: mockResults,
      count: mockResults.length,
      loading: false,
      isEmpty: false,
      requestMore: mockRequestMore,
    });

    (useInView as jest.Mock).mockReturnValue([jest.fn(), true]);

    render(<ResultsList />);

    expect(mockRequestMore).toHaveBeenCalled();
  });

  it('renders "Nothing found" text when results are empty', () => {
    (SearchContext.useSearch as jest.Mock).mockReturnValue({
      ...searchMock,
      isEmpty: true
    });

    (useInView as jest.Mock).mockReturnValue([jest.fn(), false]);

    render(<ResultsList />);

    expect(screen.getByText('Nothing found')).toBeInTheDocument();
  });
});
