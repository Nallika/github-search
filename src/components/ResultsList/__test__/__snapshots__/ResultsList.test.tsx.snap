// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsList \`<ResultsList />\` Should render component 1`] = `
<div>
  <section
    class="sc-gKsecS eogzWd"
  >
    <div
      class="sc-hKgKIp hcYfTw"
    >
      <div
        class="sc-jSguLX ialiIK"
        data-testid="loader"
      />
    </div>
  </section>
</div>
`;
